Create e2e tests for Custom Metrics - Stackdriver Adapter and HPA based on custom metrics from Stackdriver

kawych opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
- Add e2e test for Custom Metrics - Stackdriver Adapter
- Add 2e2 test for HPA based on custom metrics from Stackdriver
- Enable HorizontalPodAutoscalerUseRESTClients option

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Horizontal pod autoscaler uses REST clients through the kube-aggregator instead of the legacy client through the API server proxy.
```

kawych committed:
Create e2e test for Custom Metrics - Stackdriver Adapter

kawych committed:
Add test for HPA

kawych committed:
Enable HorizontalPodAutoscalerUseRESTClients option

kawych commented:
/assign @mwielgus 
/unassign @fabianofranz 
/unassign @smarterclayton

mwielgus commented:
/retest

kawych commented:
/assign @fgrzadkowski 
@fgrzadkowski Please review the adapter part.
/assign @MaciekPytel 
@MaciekPytel Please review the HPA part.

fgrzadkowski commented on cmd/kube-controller-manager/app/options/options_test.go:
```diff
@@ -192,15 +192,16 @@ func TestAddFlags(t *testing.T) {
+			ReconcilerSyncLoopPeriod:              metav1.Duration{Duration: 30 * time.Second},
+			Controllers:                           []string{"foo", "bar"},
+			EnableTaintManager:                    false,
+			HorizontalPodAutoscalerUseRESTClients: true,
```
Nit pick: you're setting this to the default, so probably it's not needed.

kawych replied:
It's not default value for this struct, I added this to fix failing unit test.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-deployments.go:
```diff
@@ -0,0 +1,91 @@
+/*
```
Why is it `custom_metrics-deployments.go` and not `custom_metrics_deployments.go`?

fgrzadkowski replied:
Same comment to the other file

kawych replied:
Renamed, this is by mistake.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-deployments.go:
```diff
@@ -0,0 +1,91 @@
+limitations under the License.
+*/
+
+package monitoring
```
Can you add a comment what this package contains?

kawych replied:
I'm not sure what do you expect. Do you mean 'file', or do you want me to add the same comment in all files for the package?

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-deployments.go:
```diff
@@ -0,0 +1,91 @@
+	MetricValue2     = int64(446)
+)
+
+func SDExporterDeployment(name string, replicas int32, metricValue int64) *extensions.Deployment {
```
How about using full names? StackdriverExporterDeployment. I think it'd be more readable.

kawych replied:
Changed to full names

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-deployments.go:
```diff
@@ -0,0 +1,91 @@
+	MetricValue2     = int64(446)
+)
+
+func SDExporterDeployment(name string, replicas int32, metricValue int64) *extensions.Deployment {
```
Please add a comment what this deployment will be doing.

kawych replied:
Done, same for SDExporterPod

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-deployments.go:
```diff
@@ -0,0 +1,91 @@
+		Containers: []corev1.Container{
+			{
+				Name:            "sd-exporter",
+				Image:           "gcr.io/google-containers/sd-dummy-exporter:v0.1.0",
```
Please use full name. "sd" is not well understood abbreviation.

kawych replied:
Done

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+limitations under the License.
+*/
+
+package monitoring
```
Please add a comment what this package contains.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, gcm.CloudPlatformScope)
+
+	// Hack for running tests locally
```
Is it needed to authenticate locally in Stackdriver?

kawych replied:
Yes, added a comment about that

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+	framework.ExpectNoError(err)
+
+	// Set up a cluster: create a custom metric and set up k8s-sd adapter
+	err = createDescriptors(gcmService, projectId)
```
I think that with the current implementation you will never cleanup descriptors if you leak for whatever reason the metric. Would it make sense to delete metric from in case it already exists? If you agree I'm fine with leaving it as a TODO and fixing it in a follow-up PR.

kawych replied:
Added TODO. Not sure if I understood your comment correctly, but in general I agree that we should clean up time series as well.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+func cleanupDescriptors(service *gcm.Service, projectId string) {
+	_, _ = service.Projects.MetricDescriptors.Delete(fmt.Sprintf("projects/%s/metricDescriptors/custom.googleapis.com/%s", projectId, CustomMetricName)).Do()
```
Can it return an error? If so, it'd make sense to log it for debugging purposes.

kawych replied:
Added logging.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+func cleanupSDExporterPod(cs clientset.Interface) {
+	_ = cs.Core().Pods("default").Delete("sd-exporter-1", &metav1.DeleteOptions{})
```
Same here.

kawych replied:
Done.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-deployments.go:
```diff
@@ -0,0 +1,91 @@
+var (
+	CustomMetricName = "foo-metric"
+	UnusedMetricName = "unused-metric"
+	MetricValue1     = int64(448)
```
How about `CustomMetricValue` and `UnusedMetricValue`? It'd more clear which value belongs to which metric.

kawych replied:
Changed

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+	defer cleanupSDExporterPod(kubeClient)
+
+	// Wait a short amount of time to create a pod and export some metrics
+	time.Sleep(60 * time.Second)
```
I think you should explicitly wait for pod to be schedule or metric to have some values in SD. Otherwise this test might be very flaky. WDYT?

fgrzadkowski replied:
If you agree we might fix it in the follow-up PR.

kawych replied:
I agree, this also seems to be a common pattern in integration tests.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go:
```diff
@@ -0,0 +1,200 @@
+			framework.Failf("Unexpected metric %s. Only metric %s should be supported", resource.Name, CustomMetricName)
+		}
+	}
+	value, err := customMetricsClient.NamespacedMetrics("default").GetForObject(schema.GroupKind{Group: "", Kind: "Pod"}, "sd-exporter-1", CustomMetricName)
```
Consider splitting this test into smaller methods, e.g. verifyPodMetrics, verifyPodSelectorMetrics

kawych replied:
What do you think would it improve? There is no repeated code that can be extracted. Do you think this would be more readable if each case would have it's own verify function? I prefer to leave it as it is, but I can change it if you think it would affect readability in a significant way.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,203 @@
+limitations under the License.
+*/
+
+package monitoring
```
Please add a comment about what this package contains.

fgrzadkowski commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,203 @@
+/*
```
I believe this test should be under autoscaling, not instrumentation.

kawych replied:
Moved to autoscaling

fgrzadkowski reviewed (changes requested):
I haven't reviewed autoscaling test. Assuming that @MaciekPytel will do that.

kawych commented:
@fgrzadkowski Addressed some of the comments
@MaciekPytel @mwielgus Looks like enabling HorizontalPodAutoscalerUseRESTClients option causes HPA e2e tests to fail, see https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/53205/pull-kubernetes-e2e-gce-bazel/32511/

MaciekPytel commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,205 @@
+		framework.Failf("Failed to create gcm service, %v", err)
+	}
+
+	framework.ExpectNoError(err)
```
It looks like you're checking for the same error twice

kawych replied:
Fixed.

MaciekPytel commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,205 @@
+	// Wait a for HPA to scale down targets
+	// TODO: wait for the deployments to reach the expected numbers of replicas instead of fixed
+	//       amount of time
+	time.Sleep(240 * time.Second)
```
Is there a particular reason not to implement this TODO immediately?

kawych replied:
Implemented. I left a TODO because I didn't know how much work would it be and this PR was pretty urgent (I must wait for #53330 though)

MaciekPytel commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,205 @@
+	time.Sleep(240 * time.Second)
+
+	// Verify that the deployments were scaled down to the minimum value
+	sdExporterDeployment, err := kubeClient.Extensions().Deployments("default").Get("stackdriver-exporter-deployment", metav1.GetOptions{})
```
e2e test creates a namespace for the test (f.Namespace) and I think it's prefered to use this namespace instead of 'default'.

MaciekPytel replied:
Also ideally the names of those deployments should be consts and not strings hardcoded in different places.

kawych replied:
Done

MaciekPytel commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,205 @@
+		framework.Failf("Unexpected number of replicas for 'dummy-deployment'. Expected 1, but received %v", *dummyDeployment.Spec.Replicas)
+	}
+
+	framework.ExpectNoError(err)
```
It looks like you already checked this error before in if on line 120.

kawych replied:
Removed.

MaciekPytel commented on test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,205 @@
+					},
+				},
+			},
+			MaxReplicas: 3, // default min is 1
```
Nitpicking: Wouldn't it be better to explicitly set MinReplicas here rather than comment the default happens to be right, if the test relies on it?

kawych replied:
Done.

MaciekPytel commented:
Reviewed the autoscaling part, left some comments.

kawych commented:
/retest

piosz commented:
@kawych please add release notes.

piosz commented on test/e2e/autoscaling/custom_metrics_autoscaling.go:
```diff
@@ -0,0 +1,211 @@
+
+var _ = SIGDescribe("[HPA] Horizontal pod autoscaling (scale resource: Custom Metrics from Stackdriver)", func() {
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
```
This will not work on GKE - the metrics server is not enabled there yet.

kawych replied:
Removed "gke" for now. AFAIU this affects only the autoscaling test and not the other one.

kawych commented:
Note: enabling HorizontalPodAutoscalerUseRESTClients option causes HPA e2e tests to fail. The root cause: https://github.com/kubernetes-incubator/metrics-server/issues/16
PR https://github.com/kubernetes/kubernetes/pull/53330 should fix this issue.

piosz commented:
cc @DirectXMan12

kawych commented:
/retest

fgrzadkowski commented:
LGTM for non-autoscaler part.

piosz commented:
@kawych before merging please clean up commits.

mwielgus commented:
/lgtm

piosz commented:
/approve no-issue

piosz commented:
@mikedanese could you please approve?

piosz commented:
@wojtek-t could you please approve?

wojtek-t commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -1513,6 +1513,11 @@ function start-kube-apiserver {
     params+=" --experimental-encryption-provider-config=${encryption_provider_config_path}"
   fi
 
+  # disable using HPA metrics REST clients if metrics-server isn't enabled
```
Does it work?

It seems the "horizontal-pod-autoscaler-use-rest-clients" is the controller manager flag, but here we are generating flags to apiserver...

piosz replied:
Thanks for catching this!
@kawych please fix

kawych replied:
Thanks, nice catch. Fixed & tested

kawych commented:
@mwielgus @piosz please re-add lgtm
@wojtek-t PTAL

piosz commented:
/lgtm

wojtek-t commented:
/approve no-issue

luxas commented:
@DirectXMan12 @piosz @kawych This should have the action required before upgrading release note and a description how this is a breaking change.

luxas commented:
@DirectXMan12 should we add an e2e test with your adapter as well?
So these e2e tests may be run outside of GCE? That would benefit many.

DirectXMan12 commented:
yeah.  These tests should not be blocking, if they are currently, since they depend on an external service.  I've got someone on my team (@frobware) spinning up to write HPA e2es with a dummy adapter, so that we don't have an issues with depending on any kind of external service.

DirectXMan12 commented:
@frobware you'll probably want to use the tests as a base

luxas commented:
They aren't blocking right now either, but it would be nice with a cloudprovider agnostic solution here. Thank you both!

jpbetz commented:
Should this be cherrypicked to 1.8?  If so please add `cherrypick-candidate` label.

luxas commented:
This must not be cherrypicked -- at least not in full. It bumps the default in a backwards-incompatible way.
@DirectXMan12 @piosz thoughts? I guess we are fine with master coverage

jpbetz commented:
I'm removing milestone 1.8 from this for now. Please re-add if this needs to be cherrypicked.
